File: android/src/main/kotlin/com/ble_mesh/BleMeshPlugin.kt
package com.ble_mesh

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.Log
import com.ble_mesh.models.PowerMode
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** BleMeshPlugin */
class BleMeshPlugin : FlutterPlugin, MethodCallHandler {
    private val tag = "BleMeshPlugin"

    // Method channel for communication between Flutter and native Android
    private lateinit var methodChannel: MethodChannel

    // Event channels for streaming data to Flutter
    private lateinit var messageEventChannel: EventChannel
    private lateinit var peerConnectedEventChannel: EventChannel
    private lateinit var peerDisconnectedEventChannel: EventChannel
    private lateinit var meshEventChannel: EventChannel
    private lateinit var discoveredPeersEventChannel: EventChannel

    // Event stream handlers
    private var messageStreamHandler: EventStreamHandler? = null
    private var peerConnectedStreamHandler: EventStreamHandler? = null
    private var peerDisconnectedStreamHandler: EventStreamHandler? = null
    private var meshEventStreamHandler: EventStreamHandler? = null
    private var discoveredPeersStreamHandler: EventStreamHandler? = null

    // Bluetooth mesh service
    private var meshService: BluetoothMeshService? = null
    private var context: Context? = null

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext

        // Setup method channel
        methodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "ble_mesh")
        methodChannel.setMethodCallHandler(this)

        // Setup event channels
        messageEventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "ble_mesh/messages")
        messageStreamHandler = EventStreamHandler()
        messageEventChannel.setStreamHandler(messageStreamHandler)

        peerConnectedEventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "ble_mesh/peer_connected")
        peerConnectedStreamHandler = EventStreamHandler()
        peerConnectedEventChannel.setStreamHandler(peerConnectedStreamHandler)

        peerDisconnectedEventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "ble_mesh/peer_disconnected")
        peerDisconnectedStreamHandler = EventStreamHandler()
        peerDisconnectedEventChannel.setStreamHandler(peerDisconnectedStreamHandler)

        meshEventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "ble_mesh/mesh_events")
        meshEventStreamHandler = EventStreamHandler()
        meshEventChannel.setStreamHandler(meshEventStreamHandler)

        discoveredPeersEventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "ble_mesh/discovered_peers")
        discoveredPeersStreamHandler = EventStreamHandler()
        discoveredPeersEventChannel.setStreamHandler(discoveredPeersStreamHandler)

        Log.d(tag, "BleMeshPlugin attached to engine")
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "getPlatformVersion" -> {
                result.success("Android ${android.os.Build.VERSION.RELEASE}")
            }
            "initialize" -> {
                handleInitialize(call, result)
            }
            "startMesh" -> {
                handleStartMesh(result)
            }
            "stopMesh" -> {
                handleStopMesh(result)
            }
            "sendPublicMessage" -> {
                handleSendPublicMessage(call, result)
            }
            "getConnectedPeers" -> {
                handleGetConnectedPeers(result)
            }
            "startDiscovery" -> {
                handleStartDiscovery(result)
            }
            "stopDiscovery" -> {
                handleStopDiscovery(result)
            }
            "getDiscoveredPeers" -> {
                handleGetDiscoveredPeers(result)
            }
            "connectToPeer" -> {
                handleConnectToPeer(call, result)
            }
            "disconnectFromPeer" -> {
                handleDisconnectFromPeer(call, result)
            }
            "getPeerConnectionState" -> {
                handleGetPeerConnectionState(call, result)
            }
            "blockPeer" -> {
                handleBlockPeer(call, result)
            }
            "unblockPeer" -> {
                handleUnblockPeer(call, result)
            }
            "isPeerBlocked" -> {
                handleIsPeerBlocked(call, result)
            }
            "getBlockedPeers" -> {
                handleGetBlockedPeers(result)
            }
            "clearBlocklist" -> {
                handleClearBlocklist(result)
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    private fun handleInitialize(call: MethodCall, result: Result) {
        try {
            val nickname = call.argument<String?>("nickname")
            val enableEncryption = call.argument<Boolean>("enableEncryption") ?: true
            val powerModeValue = call.argument<Int>("powerMode") ?: PowerMode.BALANCED.value
            val powerMode = PowerMode.fromInt(powerModeValue)

            // Create mesh service if not exists
            if (meshService == null && context != null) {
                meshService = BluetoothMeshService(context!!)
                setupMeshServiceCallbacks()
            }

            // Initialize mesh service
            meshService?.initialize(nickname, enableEncryption, powerMode)

            Log.d(tag, "Initialized mesh service")
            result.success(null)
        } catch (e: Exception) {
            Log.e(tag, "Error initializing mesh service", e)
            result.error("INIT_ERROR", e.message, null)
        }
    }

    private fun handleStartMesh(result: Result) {
        try {
            if (meshService == null && context != null) {
                meshService = BluetoothMeshService(context!!)
                setupMeshServiceCallbacks()
            }

            meshService?.start()
            Log.d(tag, "Started mesh service")
            result.success(null)
        } catch (e: Exception) {
            Log.e(tag, "Error starting mesh service", e)
            result.error("START_ERROR", e.message, null)
        }
    }

    private fun handleStopMesh(result: Result) {
        try {
            meshService?.stop()
            Log.d(tag, "Stopped mesh service")
            result.success(null)
        } catch (e: Exception) {
            Log.e(tag, "Error stopping mesh service", e)
            result.error("STOP_ERROR", e.message, null)
        }
    }

    private fun handleSendPublicMessage(call: MethodCall, result: Result) {
        try {
            val message = call.argument<String>("message")
            if (message == null) {
                result.error("INVALID_ARGUMENT", "Message is required", null)
                return
            }

            meshService?.sendPublicMessage(message)
            Log.d(tag, "Sent public message: $message")
            result.success(null)
        } catch (e: Exception) {
            Log.e(tag, "Error sending message", e)
            result.error("SEND_ERROR", e.message, null)
        }
    }

    private fun handleGetConnectedPeers(result: Result) {
        try {
            val peers = meshService?.getConnectedPeers() ?: emptyList()
            val peerMaps = peers.map { it.toMap() }
            result.success(peerMaps)
        } catch (e: Exception) {
            Log.e(tag, "Error getting connected peers", e)
            result.error("GET_PEERS_ERROR", e.message, null)
        }
    }

    private fun handleStartDiscovery(result: Result) {
        try {
            if (meshService == null && context != null) {
                meshService = BluetoothMeshService(context!!)
                setupMeshServiceCallbacks()
            }

            meshService?.startDiscovery()
            Log.d(tag, "Started discovery")
            result.success(null)
        } catch (e: Exception) {
            Log.e(tag, "Error starting discovery", e)
            result.error("START_DISCOVERY_ERROR", e.message, null)
        }
    }

    private fun handleStopDiscovery(result: Result) {
        try {
            meshService?.stopDiscovery()
            Log.d(tag, "Stopped discovery")
            result.success(null)
        } catch (e: Exception) {
            Log.e(tag, "Error stopping discovery", e)
            result.error("STOP_DISCOVERY_ERROR", e.message, null)
        }
    }

    private fun handleGetDiscoveredPeers(result: Result) {
        try {
            val peers = meshService?.getDiscoveredPeers() ?: emptyList()
            val peerMaps = peers.map { it.toMap() }
            result.success(peerMaps)
        } catch (e: Exception) {
            Log.e(tag, "Error getting discovered peers", e)
            result.error("GET_DISCOVERED_PEERS_ERROR", e.message, null)
        }
    }

    private fun handleConnectToPeer(call: MethodCall, result: Result) {
        try {
            val senderId = call.argument<String>("senderId")
            if (senderId == null) {
                result.error("INVALID_ARGUMENT", "senderId is required", null)
                return
            }

            val success = meshService?.connectToPeer(senderId) ?: false
            result.success(success)
        } catch (e: Exception) {
            Log.e(tag, "Error connecting to peer", e)
            result.error("CONNECT_PEER_ERROR", e.message, null)
        }
    }

    private fun handleDisconnectFromPeer(call: MethodCall, result: Result) {
        try {
            val senderId = call.argument<String>("senderId")
            if (senderId == null) {
                result.error("INVALID_ARGUMENT", "senderId is required", null)
                return
            }

            val success = meshService?.disconnectFromPeer(senderId) ?: false
            result.success(success)
        } catch (e: Exception) {
            Log.e(tag, "Error disconnecting from peer", e)
            result.error("DISCONNECT_PEER_ERROR", e.message, null)
        }
    }

    private fun handleGetPeerConnectionState(call: MethodCall, result: Result) {
        try {
            val senderId = call.argument<String>("senderId")
            if (senderId == null) {
                result.error("INVALID_ARGUMENT", "senderId is required", null)
                return
            }

            val state = meshService?.getPeerConnectionState(senderId)
            result.success(state)
        } catch (e: Exception) {
            Log.e(tag, "Error getting peer connection state", e)
            result.error("GET_PEER_STATE_ERROR", e.message, null)
        }
    }

    private fun handleBlockPeer(call: MethodCall, result: Result) {
        try {
            val senderId = call.argument<String>("senderId")
            if (senderId == null) {
                result.error("INVALID_ARGUMENT", "senderId is required", null)
                return
            }

            val success = meshService?.blockPeer(senderId) ?: false
            result.success(success)
        } catch (e: Exception) {
            Log.e(tag, "Error blocking peer", e)
            result.error("BLOCK_PEER_ERROR", e.message, null)
        }
    }

    private fun handleUnblockPeer(call: MethodCall, result: Result) {
        try {
            val senderId = call.argument<String>("senderId")
            if (senderId == null) {
                result.error("INVALID_ARGUMENT", "senderId is required", null)
                return
            }

            val success = meshService?.unblockPeer(senderId) ?: false
            result.success(success)
        } catch (e: Exception) {
            Log.e(tag, "Error unblocking peer", e)
            result.error("UNBLOCK_PEER_ERROR", e.message, null)
        }
    }

    private fun handleIsPeerBlocked(call: MethodCall, result: Result) {
        try {
            val senderId = call.argument<String>("senderId")
            if (senderId == null) {
                result.error("INVALID_ARGUMENT", "senderId is required", null)
                return
            }

            val isBlocked = meshService?.isPeerBlocked(senderId) ?: false
            result.success(isBlocked)
        } catch (e: Exception) {
            Log.e(tag, "Error checking if peer is blocked", e)
            result.error("IS_PEER_BLOCKED_ERROR", e.message, null)
        }
    }

    private fun handleGetBlockedPeers(result: Result) {
        try {
            val blockedPeers = meshService?.getBlockedPeers() ?: emptyList()
            result.success(blockedPeers)
        } catch (e: Exception) {
            Log.e(tag, "Error getting blocked peers", e)
            result.error("GET_BLOCKED_PEERS_ERROR", e.message, null)
        }
    }

    private fun handleClearBlocklist(result: Result) {
        try {
            meshService?.clearBlocklist()
            result.success(null)
        } catch (e: Exception) {
            Log.e(tag, "Error clearing blocklist", e)
            result.error("CLEAR_BLOCKLIST_ERROR", e.message, null)
        }
    }

    private fun setupMeshServiceCallbacks() {
        meshService?.onPeerDiscovered = { peer ->
            // Send discovered peer to stream
            discoveredPeersStreamHandler?.sendEvent(peer.toMap())
        }

        meshService?.onPeerConnected = { peer ->
            peerConnectedStreamHandler?.sendEvent(peer.toMap())
        }

        meshService?.onPeerDisconnected = { peer ->
            peerDisconnectedStreamHandler?.sendEvent(peer.toMap())
        }

        meshService?.onMeshEvent = { event ->
            meshEventStreamHandler?.sendEvent(event.toMap())
        }

        meshService?.onMessageReceived = { message ->
            messageStreamHandler?.sendEvent(message.toMap())
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)

        messageEventChannel.setStreamHandler(null)
        peerConnectedEventChannel.setStreamHandler(null)
        peerDisconnectedEventChannel.setStreamHandler(null)
        meshEventChannel.setStreamHandler(null)

        meshService?.cleanup()
        meshService = null
        context = null

        Log.d(tag, "BleMeshPlugin detached from engine")
    }

    /**
     * Event stream handler for sending events to Flutter
     * Ensures all events are sent on the UI thread
     */
    private class EventStreamHandler : EventChannel.StreamHandler {
        private var eventSink: EventChannel.EventSink? = null
        private val mainHandler = Handler(Looper.getMainLooper())

        override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
            eventSink = events
        }

        override fun onCancel(arguments: Any?) {
            eventSink = null
        }

        fun sendEvent(event: Any) {
            mainHandler.post {
                eventSink?.success(event)
            }
        }

        fun sendError(errorCode: String, errorMessage: String?, errorDetails: Any?) {
            mainHandler.post {
                eventSink?.error(errorCode, errorMessage, errorDetails)
            }
        }
    }
}

File: ios/Classes/BleMeshPlugin.swift
import Flutter
import UIKit

public class BleMeshPlugin: NSObject, FlutterPlugin {
    private let tag = "BleMeshPlugin"

    // Method channel for communication between Flutter and native iOS
    private var methodChannel: FlutterMethodChannel?

    // Event channels for streaming data to Flutter
    private var messageEventChannel: FlutterEventChannel?
    private var peerConnectedEventChannel: FlutterEventChannel?
    private var peerDisconnectedEventChannel: FlutterEventChannel?
    private var meshEventChannel: FlutterEventChannel?

    // Event stream handlers
    private var messageStreamHandler: EventStreamHandler?
    private var peerConnectedStreamHandler: EventStreamHandler?
    private var peerDisconnectedStreamHandler: EventStreamHandler?
    private var meshEventStreamHandler: EventStreamHandler?

    // Bluetooth mesh service
    private var meshService: BluetoothMeshService?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = BleMeshPlugin()

        // Setup method channel
        let methodChannel = FlutterMethodChannel(name: "ble_mesh", binaryMessenger: registrar.messenger())
        instance.methodChannel = methodChannel
        registrar.addMethodCallDelegate(instance, channel: methodChannel)

        // Setup event channels
        let messageEventChannel = FlutterEventChannel(name: "ble_mesh/messages", binaryMessenger: registrar.messenger())
        instance.messageStreamHandler = EventStreamHandler()
        messageEventChannel.setStreamHandler(instance.messageStreamHandler)
        instance.messageEventChannel = messageEventChannel

        let peerConnectedEventChannel = FlutterEventChannel(name: "ble_mesh/peer_connected", binaryMessenger: registrar.messenger())
        instance.peerConnectedStreamHandler = EventStreamHandler()
        peerConnectedEventChannel.setStreamHandler(instance.peerConnectedStreamHandler)
        instance.peerConnectedEventChannel = peerConnectedEventChannel

        let peerDisconnectedEventChannel = FlutterEventChannel(name: "ble_mesh/peer_disconnected", binaryMessenger: registrar.messenger())
        instance.peerDisconnectedStreamHandler = EventStreamHandler()
        peerDisconnectedEventChannel.setStreamHandler(instance.peerDisconnectedStreamHandler)
        instance.peerDisconnectedEventChannel = peerDisconnectedEventChannel

        let meshEventChannel = FlutterEventChannel(name: "ble_mesh/mesh_events", binaryMessenger: registrar.messenger())
        instance.meshEventStreamHandler = EventStreamHandler()
        meshEventChannel.setStreamHandler(instance.meshEventStreamHandler)
        instance.meshEventChannel = meshEventChannel

        print("[BleMeshPlugin] Plugin registered")
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getPlatformVersion":
            result("iOS " + UIDevice.current.systemVersion)

        case "initialize":
            handleInitialize(call, result: result)

        case "startMesh":
            handleStartMesh(result: result)

        case "stopMesh":
            handleStopMesh(result: result)

        case "sendPublicMessage":
            handleSendPublicMessage(call, result: result)

        case "sendPrivateMessage":
            handleSendPrivateMessage(call, result: result)


        case "sharePublicKey":
            handleSharePublicKey(call, result: result)
        case "getConnectedPeers":
            handleGetConnectedPeers(result: result)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func handleInitialize(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any] else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments", details: nil))
            return
        }

        let nickname = args["nickname"] as? String
        let enableEncryption = args["enableEncryption"] as? Bool ?? true
        let powerModeValue = args["powerMode"] as? Int ?? PowerMode.balanced.rawValue
        let powerMode = PowerMode.fromInt(powerModeValue)

        // Create mesh service if not exists
        if meshService == nil {
            meshService = BluetoothMeshService()
            setupMeshServiceCallbacks()
        }

        // Initialize mesh service
        meshService?.initialize(nickname: nickname, encryption: enableEncryption, mode: powerMode)

        print("[\(tag)] Initialized mesh service")
        result(nil)
    }

    private func handleStartMesh(result: @escaping FlutterResult) {
        if meshService == nil {
            meshService = BluetoothMeshService()
            setupMeshServiceCallbacks()
        }

        meshService?.start()
        print("[\(tag)] Started mesh service")
        result(nil)
    }

    private func handleStopMesh(result: @escaping FlutterResult) {
        meshService?.stop()
        print("[\(tag)] Stopped mesh service")
        result(nil)
    }

    private func handleSendPublicMessage(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let message = args["message"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Message is required", details: nil))
            return
        }

        meshService?.sendPublicMessage(content: message)
        print("[\(tag)] Sent public message: \(message)")
        result(nil)
    }

    private func handleSendPrivateMessage(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let peerId = args["peerId"] as? String,
              let encryptedData = args["encryptedData"] as? FlutterStandardTypedData,
              let senderPublicKey = args["senderPublicKey"] as? FlutterStandardTypedData else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "peerId, encryptedData, and senderPublicKey are required", details: nil))
            return
        }

        meshService?.sendPrivateMessage(peerId: peerId, encryptedData: encryptedData.data, senderPublicKey: senderPublicKey.data)
        print("[\(tag)] Sent private message to: \(peerId)")
        result(nil)
    }

    private func handleSharePublicKey(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let peerId = args["peerId"] as? String,
              let publicKey = args["publicKey"] as? FlutterStandardTypedData else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "peerId and publicKey are required", details: nil))
            return
        }

        meshService?.sharePublicKey(peerId: peerId, publicKey: publicKey.data)
        print("[\(tag)] Shared public key with: \(peerId)")
        result(nil)
    }

    private func handleGetConnectedPeers(result: @escaping FlutterResult) {
        let peers = meshService?.getConnectedPeers() ?? []
        let peerMaps = peers.map { $0.toMap() }
        result(peerMaps)
    }

    private func setupMeshServiceCallbacks() {
        meshService?.onPeerDiscovered = { [weak self] peer in
            // Peer discovered events are sent through mesh events
        }

        meshService?.onPeerConnected = { [weak self] peer in
            self?.peerConnectedStreamHandler?.sendEvent(peer.toMap())
        }

        meshService?.onPeerDisconnected = { [weak self] peer in
            self?.peerDisconnectedStreamHandler?.sendEvent(peer.toMap())
        }

        meshService?.onMeshEvent = { [weak self] event in
            self?.meshEventStreamHandler?.sendEvent(event.toMap())
        }

        meshService?.onMessageReceived = { [weak self] message in
            self?.messageStreamHandler?.sendEvent(message.toMap())
        }
    }

    public func detachFromEngine(for registrar: FlutterPluginRegistrar) {
        methodChannel?.setMethodCallHandler(nil)
        messageEventChannel?.setStreamHandler(nil)
        peerConnectedEventChannel?.setStreamHandler(nil)
        peerDisconnectedEventChannel?.setStreamHandler(nil)
        meshEventChannel?.setStreamHandler(nil)

        meshService?.cleanup()
        meshService = nil

        print("[\(tag)] Plugin detached from engine")
    }
}

/// Event stream handler for sending events to Flutter
/// Ensures all events are sent on the main thread
class EventStreamHandler: NSObject, FlutterStreamHandler {
    private var eventSink: FlutterEventSink?

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }

    func sendEvent(_ event: Any) {
        DispatchQueue.main.async { [weak self] in
            self?.eventSink?(event)
        }
    }

    func sendError(code: String, message: String?, details: Any?) {
        DispatchQueue.main.async { [weak self] in
            self?.eventSink?(FlutterError(code: code, message: message, details: details))
        }
    }
}
